src/mono/mono/metadata/reflection.c

```diff
@@ -2008,450 +2138,140 @@ mono_reflection_parse_type_checked (char *name, MonoTypeNameParse *info, MonoErr
-
-return_null:
-	type = NULL;
-	goto exit;
-
-exit:
-	if (fullName)
-		g_string_free (fullName, TRUE);
+leave:
 	HANDLE_FUNCTION_RETURN_VAL (type);
 }
 
-/**
- * mono_reflection_free_type_info:
- */
-void
-mono_reflection_free_type_info (MonoTypeNameParse *info)
-{
-	g_list_free (info->modifiers);
-	g_list_free (info->nested);
-
-	if (info->type_arguments) {
-		for (guint i = 0; i < info->type_arguments->len; i++) {
-			MonoTypeNameParse *subinfo = (MonoTypeNameParse *)g_ptr_array_index (info->type_arguments, i);
-
-			mono_reflection_free_type_info (subinfo);
-			/*We free the subinfo since it is allocated by _mono_reflection_parse_type*/
-			g_free (subinfo);
-		}
-
-		g_ptr_array_free (info->type_arguments, TRUE);
-	}
-}
-
-/**
- * mono_reflection_type_from_name:
- * \param name type name.
- * \param image a metadata context (can be NULL).
- *
- * Retrieves a \c MonoType from its \p name. If the name is not fully qualified,
- * it defaults to get the type from \p image or, if \p image is NULL or loading
- * from it fails, uses corlib.
- *
- */
-MonoType*
-mono_reflection_type_from_name (char *name, MonoImage *image)
-{
-	MonoType *result;
-	MONO_ENTER_GC_UNSAFE;
-	ERROR_DECL (error);
-
-	MonoAssemblyLoadContext *alc = mono_alc_get_default ();
-
-	result = mono_reflection_type_from_name_checked (name, alc, image, error);
-
-	mono_error_cleanup (error);
-	MONO_EXIT_GC_UNSAFE;
-	return result;
-}
-
 /**
  * mono_reflection_type_from_name_checked:
  * \param name type name.
  * \param alc the AssemblyLoadContext to check/load into
  * \param image a metadata context (can be NULL).
+ * \param ignorecase compare type names case-insensitively
+ * \param use_toplevel_assembly if true, follow the semantics of Assembly.GetAssembly
  * \param error set on error.
  * Retrieves a MonoType from its \p name. If the name is not fully qualified,
  * it defaults to get the type from \p image or, if \p image is NULL or loading
  * from it fails, uses corlib.  On failure returns NULL and sets \p error.
  */
 MonoType*
-mono_reflection_type_from_name_checked (char *name, MonoAssemblyLoadContext *alc, MonoImage *image, MonoError *error)
+mono_reflection_type_from_name_checked (char *name, MonoAssemblyLoadContext *alc, MonoImage *image, gboolean ignorecase, gboolean use_toplevel_assembly, MonoError *error)
 {
-	MonoType *type = NULL;
-	MonoTypeNameParse info;
-	char *tmp;
-
-	error_init (error);
-	/* Make a copy since parse_type modifies its argument */
-	tmp = g_strdup (name);
-
-	/*g_print ("requested type %s\n", str);*/
-	ERROR_DECL (parse_error);
-	if (!mono_reflection_parse_type_checked (tmp, &info, parse_error)) {
-		mono_error_cleanup (parse_error);
-		goto leave;
-	}
-	type = _mono_reflection_get_type_from_info (alc, &info, image, FALSE, TRUE, error);
-leave:
-	g_free (tmp);
-	mono_reflection_free_type_info (&info);
-	return type;
+	return mono_reflection_type_from_name_internal (name, alc, image, ignorecase, use_toplevel_assembly, NULL, error);
```

Comment:
```suggestion
	gboolean type_resolve = FALSE;
	return mono_reflection_type_from_name_internal (name, alc, image, ignorecase, use_toplevel_assembly, type_resolve, error);
```

The function used to call `_mono_reflection_get_type_from_info` which sets `type_resolve` to false.